### ISSUE TITLE
Misleading TAIT diagnostic on missing `Send + Sync` bound

### ISSUE THREAD

**Issue by ivan770** [rust-lang/rust#99799]
I tried this code:

https://gist.github.com/ivan770/709330e60ce2dcfbfcb04d5dd539247b
(I think it's possible to minimize it a bit, however I'm not sure on how to do that in this case)

I expected to see this happen: It provides a meaningful diagnostic that I have to add `Send + Sync` bound to `NonSendSync`.

Instead, this happened: Misleading error occures:
```
error[E0391]: cycle detected when type-checking `<impl at src/lib.rs:16:1: 16:35>::call`
  --> src/lib.rs:20:10
   |
20 |     ) -> TestFut<'yielder> {
   |          ^^^^^^^^^^^^^^^^^
   |
   = note: ...which requires evaluating trait selection obligation `for<'r, 's, 't0, 't1, 't2, 't3, 't4, 't5, 't6, 't7, 't8> {core::future::ResumeTy, NonSendSync, core::pin::Pin<&'r mut NonSendSync>, &'s mut enstream::Yielder<'t0, ()>, enstream::Yielder<'t1, ()>, (), impl for<'t2, 't3, 't4> core::future::future::Future<Output = ()>, &'t5 mut core::pin::Pin<&'t6 mut NonSendSync>, futures_util::stream::stream::next::Next<'t7, core::pin::Pin<&'t8 mut NonSendSync>>}: core::marker::Send`...
note: ...which requires computing type of `NonSendSync::{opaque#0}`...
  --> src/lib.rs:34:20
   |
34 | type NonSendSync = impl Stream<Item = ()>;
   |                    ^^^^^^^^^^^^^^^^^^^^^^
   = note: ...which again requires type-checking `<impl at src/lib.rs:16:1: 16:35>::call`, completing the cycle
note: cycle used when computing type of `TestFut::{opaque#0}`
  --> src/lib.rs:10:26
   |
10 | type TestFut<'yielder> = impl Future<Output = ()> + Send + Sync;
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0391`.
```

### Meta

`rustc --version --verbose`:
```
rustc 1.64.0-nightly (6dbae3ad1 2022-07-25)
binary: rustc
commit-hash: 6dbae3ad19309bb541d9e76638e6aa4b5449f29a
commit-date: 2022-07-25
host: aarch64-unknown-linux-musl
release: 1.64.0-nightly
LLVM version: 14.0.6
```

</p>
</details>

**Comment by ivan770** [comment]
Duplicate of #55997
